### PR TITLE
IndexTranslator

### PR DESCRIPTION
**Pull request by SimonHeybrock**
@OwenArnold This is a first attempt at what we discussed last week. The test `test_split_2_to_3_combine_results` shows how I imagine this being used to extract data and deal with MPI. It could benefit from some helper functions, but have a look first if this is roughly ok.

### CODE REVIEW

File: cow_instrument/testing/IndexTranslatorTest.cpp [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+
+    // 2. Use indices to get data from object that should be redistributed. Here
+    // we simply keep indices as fake data.
+    const auto &data = indices;
+
+    // 3. Tag data with GlobalSpectrumIndex and SpectrumNumber
+    auto globalIndices = t[partition].globalSpectrumIndices(p3);
+    auto spectrumNumbers = t[partition].spectrumNumbers(p3);
+    std::vector<
+        std::vector<std::tuple<GlobalSpectrumIndex, SpectrumNumber, Data>>>
```
Comment (SimonHeybrock) [inline cow_instrument/testing/IndexTranslatorTest.cpp]: Including spectrum numbers might be unnecessary. If we store the complete list of spectrum numbers in the translator having the global spectrum index should be enough.

File: cow_instrument/testing/IndexTranslatorTest.cpp [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+      for (const auto &item : partitionData)
+        result.push_back(item);
+
+    // 6. Sort data according to global spectrum index
+    std::sort(
+        result.begin(), result.end(),
+        [](const std::tuple<GlobalSpectrumIndex, SpectrumNumber, Data> &t1,
+           const std::tuple<GlobalSpectrumIndex, SpectrumNumber, Data> &t2)
+            -> bool { return std::get<0>(t1) < std::get<0>(t2); });
+
```
Comment (SimonHeybrock) [inline cow_instrument/testing/IndexTranslatorTest.cpp]: We now have vector of data in the correct order, i.e., the index into the vector is the spectrum index with the new partitioning. We could now use that vector to build the new instrument or `SpectrumInfo`.

File: cow_instrument/testing/IndexTranslatorTest.cpp (hunk 1 above)
Comment (OwenArnold) [inline cow_instrument/testing/IndexTranslatorTest.cpp]: Agreed. But I suggest we leave it in the prototype for the time being.

File: cow_instrument/testing/IndexTranslatorTest.cpp [hunk 3]
```diff
@@ -9,7 +9,39 @@ using namespace testing;
...
+      << "Partition should always be zero";
+  ;
+  globalSpectrumIndex = 11;
+  EXPECT_EQ(0, p.partitionIndexOf(globalSpectrumIndex))
+      << "Partition should always be zero";
+  ;
+}
+
+TEST(index_translator_test, test_simple_index_translator_partitioning) {
+  Partitioning singlePartition(1 /*Two partitions*/);
```
Comment (SimonHeybrock) [inline cow_instrument/testing/IndexTranslatorTest.cpp]: `2`?

Comment (OwenArnold) [inline cow_instrument/testing/IndexTranslatorTest.cpp]: My bad. I'll remove that.

Comment (OwenArnold) [inline cow_instrument/testing/IndexTranslatorTest.cpp]: Why is the branch name starting with 27?

Comment (SimonHeybrock) [inline cow_instrument/testing/IndexTranslatorTest.cpp]: It is the issue index.

**Comment by OwenArnold** [comment]
@SimonHeybrock. This is very useful. Particularly the nice tests you added. I suggest we merge this.

File: cow_instrument/testing/IndexTranslatorTest.cpp (hunk 3 above)
Comment (OwenArnold) [inline cow_instrument/testing/IndexTranslatorTest.cpp]: OK.

**Comment by SimonHeybrock** [comment]
Great, let me know if you need more functionality from the translator.

**Comment by OwenArnold** [comment]
It's a good stand-alone start. I don't think there's any point making this more complicated until I'm able to use it directly.

**Pull request merged**